### PR TITLE
Replace reflection usage with MethodHandles when performance matters

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

As we have java8 as a minimum target we can use MethodHandles. We should do so when we expect to have a method called multiple times.

Modifications:

- Replace usage of reflection with MethodHandles where it makes sense
- Remove some code which was there to support java < 8

Result:

Faster code

### CODE REVIEW

**Comment by normanmaurer** [comment]
We could also use `MethodHandle` in 4.1 if we are running on Java8+

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java [hunk 1]
```diff
@@ -30,46 +32,53 @@
...
                 try {
+                    MethodHandles.Lookup lookup = MethodHandles.lookup();
                     // See https://bugs.openjdk.java.net/browse/JDK-8171377
-                    Method m = PlatformDependent0.UNSAFE.getClass().getDeclaredMethod(
-                            "invokeCleaner", ByteBuffer.class);
-                    m.invoke(PlatformDependent0.UNSAFE, buffer);
+                    MethodHandle m = lookup.findVirtual(
+                            PlatformDependent0.UNSAFE.getClass(),
+                            "invokeCleaner",
+                            MethodType.methodType(void.class).appendParameterTypes(ByteBuffer.class)).
```
Comment (carl-mastrangelo) [inline common/src/main/java/io/netty/util/internal/CleanerJava9.java]: `MethodType.methodType(void.class, ByteBuffer.class)`

?

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java [hunk 2]
```diff
@@ -78,7 +87,7 @@ public void freeDirectBuffer(ByteBuffer buffer) {
         // See https://bugs.openjdk.java.net/browse/JDK-8191053.
         if (System.getSecurityManager() == null) {
             try {
-                INVOKE_CLEANER.invoke(PlatformDependent0.UNSAFE, buffer);
+                INVOKE_CLEANER_HANDLE.invokeExact(buffer);
```
Comment (carl-mastrangelo) [inline common/src/main/java/io/netty/util/internal/CleanerJava9.java]: In the past I have had pain using invokeExact (if the types aren't exactly right, normal covariance doesn't work, or something like that).  

Consider using plain `invoke`.   I'm moderately confident the "exact" overload doesn't have any performance benefit.

File: handler/src/main/java/io/netty/handler/ssl/Conscrypt.java [hunk 3]
```diff
@@ -16,23 +16,27 @@
...
-    private static Method loadIsConscryptEngine() {
+    private static MethodHandle loadIsConscryptEngine() {
         try {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+
             Class<?> conscryptClass = Class.forName("org.conscrypt.Conscrypt", true,
                     ConscryptAlpnSslEngine.class.getClassLoader());
-            return conscryptClass.getMethod("isConscrypt", SSLEngine.class);
+            return lookup.findStatic(conscryptClass, "isConscrypt", MethodType.methodType(boolean.class)
+                    .appendParameterTypes(SSLEngine.class));
```
Comment (carl-mastrangelo) [inline handler/src/main/java/io/netty/handler/ssl/Conscrypt.java]: Similar comment, just use the vararg form?

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java (hunk 2 above)
Comment (franz1981) [reply common/src/main/java/io/netty/util/internal/CleanerJava9.java]: Tomorrow I will give a better look (still recovering from a fever today), but I have experienced the opposite ie invokeExact to get a huge benefit in perf if compared with just invoke. I need to re-check, but that was my past experience with it.

Comment (normanmaurer) [reply common/src/main/java/io/netty/util/internal/CleanerJava9.java]: @carl-mastrangelo I tested this one and it works... I would prefer to use `invokeExact` if possible.

File: common/src/main/java/io/netty/util/internal/CleanerJava9.java (hunk 1 above)
Comment (normanmaurer) [reply common/src/main/java/io/netty/util/internal/CleanerJava9.java]: yes... was not aware there is such an overload :D

File: codec/src/main/java/io/netty/handler/codec/compression/ByteBufChecksum.java [hunk 4]
```diff
@@ -33,43 +32,34 @@
...
+                @Override
+                public void update(ByteBuffer b) {
+                    checksum.update(b);
+                }
+            };
         }
-        if (checksum instanceof CRC32 && CRC32_UPDATE_METHOD != null) {
-            return new ReflectiveByteBufChecksum(checksum, CRC32_UPDATE_METHOD);
+        if (checksum instanceof CRC32) {
+            return new OptimizedByteBufChecksum<CRC32>((CRC32) checksum) {
```
Comment (franz1981) [inline codec/src/main/java/io/netty/handler/codec/compression/ByteBufChecksum.java]: I would create a  `CRC32 castedChecksum = (CRC32) checksum` to be used on 
```java
                @Override
                public void update(ByteBuffer b) {
                    castedChecksum.update(b);
                }
```

Comment (franz1981) [reply codec/src/main/java/io/netty/handler/codec/compression/ByteBufChecksum.java]: Same for `Adler32`: IIRC captured final are `true finals` 4 the JVM, so I would expect `checksum.update(b);` to be monomorphic, but to reduce the chance that the JVM won't be able to do that, this change would help...

Comment (normanmaurer) [reply codec/src/main/java/io/netty/handler/codec/compression/ByteBufChecksum.java]: I would prefer to keep it as it is as its cleaner imho and I am not convinced it matters.

Comment (franz1981) [reply codec/src/main/java/io/netty/handler/codec/compression/ByteBufChecksum.java]: Yep,. probably it won't worth doing it

**Comment by normanmaurer** [comment]
Nevermind we can not do this in 4.1:

```
[ERROR] ~/workspace/netty/common/src/main/java/io/netty/util/internal/CleanerJava9.java:[108,53] error: polymorphic signature calls are not supported in -target 1.6
```

**Comment by franz1981** [comment]
Ah right, was forgetting it...same problem for JCtools :O Nitsan mentioned that maybe there is a solution for that, but TBH I haven't asked what was it

**Comment by normanmaurer** [comment]
@franz1981 not a big deal... I just hoped we could maybe at least do it on the cleaners as these are often a hot path.

**Comment by normanmaurer** [comment]
@franz1981 PTAL again and approve if all looks good...

**Pull request merged**